### PR TITLE
doc(framework): pytorch guide

### PR DESCRIPTION
**Pull request by bojiang**
## What does this PR address?

Fixes #(issue)

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
  those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

## Who can help review?

Feel free to tag members/contributors who can help review your PR.

### CODE REVIEW

**Comment by aarnphm** [comment]
oo you probably need to rebase this ;)

**Comment by bojiang** [comment]
@aarnphm thx. pushed by mistake

**Comment by aarnphm** [comment]
is this ready yet?

File: docs/source/frameworks/pytorch.rst [hunk 1]
```diff
@@ -2,55 +2,158 @@
...
+Preface
+-------
+
+User may also convert PyTorch models to torchscript models, and use ``bentoml.torchscript`` 
+to save them. It will be a more stable way to serve and deploy PyTorch models in production.
+
+.. note::
+
+    :bdg-info:`Remarks:` We recommend users apply model optimization techniques such as **distillation** or **quantization**.
+    Alternatively, PyTorch models can also be converted to :ref:`ONNX <frameworks/onnx>` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.).
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:16]:
```suggestion
    :bdg-info:`Remarks:` We recommend users to apply model optimization techniques such as `distillation <https://arxiv.org/abs/1503.02531>`_ or `quantization <https://pytorch.org/docs/stable/quantization.html#general-quantization-flow>`_ .
    Alternatively, PyTorch models can also be converted to :doc:`/frameworks/onnx` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.).
```

Comment (aarnphm) [reply docs/source/frameworks/pytorch.rst:16]: add something like:
`in this tutorial, we will go over how to train a simple cnn with pytorch and use bentoml ...`

File: docs/source/frameworks/pytorch.rst [hunk 2]
```diff
@@ -2,55 +2,158 @@
...
-        self.linear2 = nn.Linear(128, vocab_size)
 
-    def forward(self, inputs):
-        embeds = self.embeddings(inputs).view((1, -1))
-        out = F.relu(self.linear1(embeds))
-        out = self.linear2(out)
-        log_probs = F.log_softmax(out, dim=1)
-        return log_probs
+    model = Net()
+    ... # training
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:54]: add reproducible training code here.

also add a link to the prior tutorial if applicable.

File: docs/source/frameworks/pytorch.rst [hunk 3]
```diff
@@ -2,55 +2,158 @@
...
-    runner = bentoml.pytorch.get(tag).to_runner()
-    runner.init_local()
-    runner.run(pd.DataFrame("/path/to/csv"))
+
+    class Net(nn.Module):
+        def __init__(self):
+            super().__init__()
+
+        def forward(self, x, y):
+            return x + y
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:77]: makes this into two sections with `1, 2` marking.

File: docs/source/frameworks/pytorch.rst [hunk 4]
```diff
@@ -2,55 +2,158 @@
...
 .. note::
 
-   You can find more examples for **PyTorch** in our `gallery <https://github.com/bentoml/gallery>`_ repo.
+    :bdg-info:`Remarks:` All of the external python modules/util functions that the model
+    requires must be included in the project. The PyTorch framework requires them to restore
+    the model.
+
+.. note::
+
+    :code:`bentoml.pytorcpytorch.save_model` has parameter ``signatures``.
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:93]:
```suggestion
    :code:`bentoml.pytorch.save_model` has parameter ``signatures``.
```

File: docs/source/frameworks/pytorch.rst [hunk 5]
```diff
@@ -2,55 +2,158 @@
...
+    class Net(nn.Module):
+        def __init__(self):
+            super().__init__()
+
+        def forward(self, x, y):
+            return x + y
+
+
+    model = Net()
+    ... # training
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:81]: same here

File: docs/source/frameworks/pytorch.rst [hunk 6]
```diff
@@ -2,55 +2,158 @@
...
+
+    bentoml.pytorch.save(model, "my_torch_model")
 
 .. note::
 
-   You can find more examples for **PyTorch** in our `gallery <https://github.com/bentoml/gallery>`_ repo.
+    :bdg-info:`Remarks:` All of the external python modules/util functions that the model
+    requires must be included in the project. The PyTorch framework requires them to restore
+    the model.
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:89]: makes this a warning. 

Also add an example how people should add the modules / utility functions to bento via `bentofile.yaml`.

You can use the YOLO example here.

**Review by aarnphm — COMMENTED** [review]
still need a lot of more docs. :(

File: docs/source/frameworks/pytorch.rst [hunk 7]
```diff
@@ -2,55 +2,158 @@
 PyTorch
 =======
 
-Here's an example using PyTorch with BentoML:
+BentoML provides native support for serving and deploying models trained from PyTorch.
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply docs/source/frameworks/pytorch.rst:5]: Add lowest version supported is 1.3.0

File: docs/source/frameworks/pytorch.rst [hunk 8]
```diff
@@ -2,55 +2,193 @@
...
+            self.fc3 = nn.Linear(84, 10)
+
+        def forward(self, x):
+            x = self.pool(F.relu(self.conv1(x)))
+            x = self.pool(F.relu(self.conv2(x)))
+            x = x.view(-1, 16 * 5 * 5)
+            x = F.relu(self.fc1(x))
+            x = F.relu(self.fc2(x))
+            x = self.fc3(x)
+            return x
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:70]: IMO we should mimic how people actually do this in production, where they often separate a model file to a separate file. 

We then can add a warning like:
```
You can then add the following files to `bentofile.yaml`: ...
```

**Review by aarnphm — COMMENTED** [review]
I can take this over if needed.

File: docs/source/frameworks/pytorch.rst [hunk 9]
```diff
@@ -2,55 +2,193 @@
...
 =======
 
-Here's an example using PyTorch with BentoML:
+BentoML provides native support for serving and deploying models trained from PyTorch. For more in-depth tutorials about PyTorch, please visit `PyTorch's official documentation <https://pytorch.org/tutorials/>`_
 
-.. code:: python
+Preface
+-------
+
+If you have already compiled your PyTorch model to TorchScript, you might consider to use :doc:`bentoml.torchscript </reference/frameworks/torchscript>`. BentoML provides first-class support for TorchScript, and hence using ``bentoml.torchscript`` are less prone to bugs during production.
```
Comment (ssheng) [inline docs/source/frameworks/pytorch.rst:10]:
```suggestion
If you have already compiled your PyTorch model to TorchScript, you might consider to use :doc:`bentoml.torchscript </reference/frameworks/torchscript>`. BentoML provides first-class support for TorchScript, hence using ``bentoml.torchscript`` is less prone to compatibility issues during production.
```

File: docs/source/frameworks/pytorch.rst [hunk 10]
```diff
@@ -2,55 +2,193 @@
...
+
+.. note::
+
+    :bdg-info:`Remarks:` We recommend users to apply model optimization techniques such as `distillation <https://arxiv.org/abs/1503.02531>`_ or `quantization <https://pytorch.org/docs/stable/quantization.html#general-quantization-flow>`_ .
+    Alternatively, PyTorch models can also be converted to :doc:`/frameworks/onnx` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.).
```
Comment (ssheng) [inline docs/source/frameworks/pytorch.rst:15]:
```suggestion
    :bdg-info:`Remarks:` We recommend users to apply model optimization techniques such as `distillation <https://arxiv.org/abs/1503.02531>`_ or `quantization <https://pytorch.org/docs/stable/quantization.html#general-quantization-flow>`_ . Alternatively, PyTorch models can also be converted to :doc:`/frameworks/onnx` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.) for better performance.
```

**Review by aarnphm — APPROVED** [review]

**Review by aarnphm — APPROVED** [review]

File: docs/source/frameworks/pytorch.rst [hunk 11]
```diff
@@ -2,55 +2,193 @@
...
+
+
+    model = Net()
+
+    import torch.optim as optim
+
+    criterion = nn.CrossEntropyLoss()
+    optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)
+
+    for epoch in range(2):
```
Comment (ssheng) [inline docs/source/frameworks/pytorch.rst:80]: Let's add a comment stating the small epoch is just for demonstration purpose.

File: docs/source/frameworks/pytorch.rst [hunk 12]
```diff
@@ -2,55 +2,193 @@
...
-    runner.run(pd.DataFrame("/path/to/csv"))
+    :bdg-info:`Remarks:` All of the external python modules/util functions that the model
+    requires must be included in the project. The PyTorch framework requires them to restore
+    the model.
 
 .. note::
 
-   You can find more examples for **PyTorch** in our `bentoml/examples https://github.com/bentoml/BentoML/tree/main/examples`_ directory.
+    :code:`bentoml.pytorch.save_model` has parameter ``signatures``.
+    ``signatures`` refers to a general :ref:`Model Signatures <concepts/model:Model Signatures>`_ that dictates which methods can be used for inference in the Runner context. This signatures dictionary will be used during the creation process of a Runner instance.
```
Comment (ssheng) [inline docs/source/frameworks/pytorch.rst:141]:
```suggestion
    The ``signatures`` argument of type :ref:`Model Signatures <concepts/model:Model Signatures>`_ in :code:`bentoml.pytorch.save_model` is used to express which methods should be used for inference and exposed in the Runner. This signatures dictionary will be used during the creation process of a Runner instance.
```

File: docs/source/frameworks/pytorch.rst [hunk 13]
```diff
@@ -2,55 +2,193 @@
...
+    async def predict(json_obj: JSONSerializable) -> JSONSerializable:
+        batch_ret = await runner.async_run([json_obj])
+        return batch_ret[0]
 
+.. note::
+
+    Follow the steps to get the best performance out of your PyTorch model.
+    #. Apply adaptive batching if possible.
+    #. Serve on GPUs if applicable.
+    #. See performance guide from [PyTorch Doc]
```
Comment (ssheng) [inline docs/source/frameworks/pytorch.rst:175]: Please link the docment.

**Comment by bojiang** [comment]
@ssheng thanks! just updated

File: docs/source/frameworks/pytorch.rst [hunk 14]
```diff
@@ -2,55 +2,192 @@
...
-    runner.run(pd.DataFrame("/path/to/csv"))
+    :bdg-info:`Remarks:` All of the external python modules/util functions that the model
+    requires must be included in the project. The PyTorch framework requires them to restore
+    the model.
 
 .. note::
 
-   You can find more examples for **PyTorch** in our `bentoml/examples https://github.com/bentoml/BentoML/tree/main/examples`_ directory.
+    :code:`bentoml.pytorch.save_model` has parameter ``signatures``.
+    The ``signatures`` argument of type :ref:`Model Signatures <concepts/model:Model Signatures>`_ in :code:`bentoml.pytorch.save_model` is used to express which methods should be used for inference and exposed in the Runner. This signatures dictionary will be used during the creation process of a Runner instance.
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:140]:
```suggestion
    The ``signatures`` argument of type :ref:`Model Signatures <concepts/model:Model Signatures>` in :obj:`bentoml.pytorch.save_model` is used to determine which methods will be used for inference and exposed in the Runner. The signatures dictionary will then be used during the creation process of a Runner instance.
```

File: docs/source/frameworks/pytorch.rst [hunk 15]
```diff
@@ -2,55 +2,192 @@
...
+argument with the method name and providing :code:`batchable` and :code:`batch_dim` configurations when saving the model to the model store.
+
+.. seealso::
+
+   See :ref:`Adaptive Batching <guides/batching:Adaptive Batching>` to learn more.
+
+
+.. note::
+
+   You can find more examples for **PyTorch** in our `bentoml/examples https://github.com/bentoml/BentoML/tree/main/examples`_ directory.
```
Comment (aarnphm) [inline docs/source/frameworks/pytorch.rst:191]:
```suggestion
   You can find more examples for **PyTorch** in our `bentoml/examples <https://github.com/bentoml/BentoML/tree/main/examples>`_ directory.
```

**Review by aarnphm — APPROVED** [review]

**Pull request merged**